drivers/adc-dac/ad5592r/ad5593r.c

```diff
@@ -122,6 +123,40 @@ int32_t ad5593r_read_adc(struct ad5592r_dev *dev, uint8_t chan,
 	return 0;
 }
 
+/**
+ * Read ADC sequence.
+ *
+ * @param dev - The device structure.
+ * @return 0 in case of success, negative error code otherwise
+ */
+int32_t ad5593r_read_adc_sequence(struct ad5592r_dev *dev,
```

Comment:
this does not look different than just reading each individual channel with ad5593r_read_adc

If a sequence of adc would want to be read together then the actual i2c transmission would be extended with the number of channels. The transfer would be 2 * NO_CHANNELS not just 2 and using a for loop.

Check Figure 37. I2c Block Read page 23 of the datasheet.

Reply:
In this case I am initiating the ADC sequence at the application level and then the ad5593r_read_adc_seqeunce function gathers the I2C frames (or SPI frames for the AD5592R) for the number ADC channels that were included in the sequence. These are added to the adc_values array returning only the adc_code. This function was added to get the data from each channel in a sequence. I could only get the data from the last channel read in the sequence with the ad5593r_read_adc function, unless there is another way to do so that I didn't think of?

Reply:
my suggestion here was to increase the transmission size to more than 2 bytes. That way you eliminate the need for the for loop. something like:
```
uint8_t data[2 * adc_seq_channels];

ret = i2c_read(dev->i2c, data, ARRAY_SIZE(data), STOP_BIT);
if (ret < 0)
      return ret;
```
This way you get all the channels data that are enabled in the sequence in one transmission. 
Then you would just use a loop to interpret the data
```
for (uint8_t i = 0; i <= adc_seq_channels; i++) 
         adc_values[i] = ((uint16_t)(((data[2 * i] & 0x0F) << 8) + data[2 * i + 1]) & 0xFFF);
```

Reply:
Looking into the above solution, it will work to read back all of the channels in a single ADC read. However, I was looking to try and enable the repeat bit in the ADC sequencer and capture the data coming from the bus indefinitely. This is what the original solution was trying to accomplish, having set n channels, read back n frames, loop at the application layer and repeat reading data. In the meantime, another solution came to mind running sections depending on seq_status, this would be an alternative to ad5593r_read_adc: Thoughts?

`int32_t ad5593r_read_adc(struct ad5592r_dev *dev,
				  uint16_t *adc_values, int32_t seq_status)
{
	uint16_t temp;
	int32_t ret;
	uint8_t i, num_ch_in_seq = 0, data[MAX_ARRAY_SIZE];

	if (!dev)
		return FAILURE;
	
	for (i = 0; i <= dev->num_channels; i++)
	{
		if (AD5592R_REG_ADC_SEQ_INCL(i) & dev->adc_sequence_reg)
			num_ch_in_seq++; 	
	}
	
	switch (seq_status)
	{
	case REPEAT_SEQUENCE:
		ret = i2c_read(dev->i2c, data, (2 * num_ch_in_seq), RESTART_BIT);
		if (ret < 0)
			return ret;
		break;
		
	case STOP_SEQUENCE:
		ret = i2c_read(dev->i2c, data, (2 * num_ch_in_seq), STOP_BIT);
		if (ret < 0)
			return ret;
		break;
		
	case START_SEQUENCE:
	default:
		/* fall-through */
		temp = dev->adc_sequence_reg;

		data[0] = AD5593R_MODE_CONF | AD5592R_REG_ADC_SEQ | (AD5592R_REG_ADC_SEQ_REP & dev->adc_sequence_reg);
		data[1] = (dev->adc_sequence_reg & 0xFF);
		data[2] = temp & 0xFF;

		ret = i2c_write(dev->i2c, data, 3, STOP_BIT);
		if (ret < 0)
			return ret;

		data[0] = AD5593R_MODE_ADC_READBACK;
		ret = i2c_write(dev->i2c, data, 1, RESTART_BIT);
		if (ret < 0)
			return ret;
		
		ret = i2c_read(dev->i2c, data, (2 * num_ch_in_seq), 
			(dev->adc_sequence_reg & AD5592R_REG_ADC_SEQ_REP) ? RESTART_BIT : STOP_BIT );
		if (ret < 0)
			return ret;
		break;
	}
	
	for (i = 0; i < num_ch_in_seq; i++)
	{
		adc_values[i] = (uint16_t)(((data[2 * i] & 0xFF) << 8) + data[(2 * i) + 1]);	
	}
	
	return 0;
}`